File: python/features/steps/list_format_steps.py
"""Behave steps for list formatting scenarios."""

from __future__ import annotations

import os

from behave import given, then, when

from features.steps.shared import (
    build_issue,
    load_project_directory,
    read_issue_file,
    write_issue_file,
)
from kanbus.config_loader import load_project_configuration
from kanbus.issue_line import compute_widths, format_issue_line
from kanbus.project import get_configuration_path


@given("issues for list color coverage exist")
def given_issues_for_list_color_coverage(context: object) -> None:
    project_dir = load_project_directory(context)
    issues = [
        build_issue("kanbus-line-epic", "Epic", "epic", "open", None, []),
        build_issue(
            "kanbus-line-task", "Task", "task", "in_progress", "kanbus-line-epic", []
        ),
        build_issue("kanbus-line-bug", "Bug", "bug", "blocked", None, []),
        build_issue("kanbus-line-story", "Story", "story", "closed", None, []),
        build_issue("kanbus-line-chore", "Chore", "chore", "backlog", None, []),
        build_issue(
            "kanbus-line-initiative", "Initiative", "initiative", "unknown", None, []
        ),
        build_issue(
            "kanbus-line-sub", "Sub", "sub-task", "open", "kanbus-line-epic", []
        ),
        build_issue("kanbus-line-event", "Event", "event", "open", None, []),
        build_issue("kanbus-line-unknown", "Unknown", "mystery", "open", None, []),
    ]
    updates = {
        "kanbus-line-epic": {"priority": 0},
        "kanbus-line-task": {"priority": 1},
        "kanbus-line-bug": {"priority": 2},
        "kanbus-line-story": {"priority": 3},
        "kanbus-line-chore": {"priority": 4},
        "kanbus-line-initiative": {"priority": 9},
        "kanbus-line-sub": {"priority": 2},
        "kanbus-line-event": {"priority": 2},
        "kanbus-line-unknown": {"priority": 2},
    }
    identifiers = []
    for issue in issues:
        issue = issue.model_copy(update=updates.get(issue.identifier, {}))
        write_issue_file(project_dir, issue)
        identifiers.append(issue.identifier)
    context.list_color_issue_ids = identifiers


@when("I format list lines for color coverage")
def when_format_list_lines_for_color_coverage(context: object) -> None:
    if not hasattr(context, "original_no_color"):
        context.original_no_color = os.environ.get("NO_COLOR")
    os.environ.pop("NO_COLOR", None)
    project_dir = load_project_directory(context)
    configuration = load_project_configuration(get_configuration_path(project_dir))
    identifiers = getattr(context, "list_color_issue_ids", [])
    issues = [read_issue_file(project_dir, identifier) for identifier in identifiers]
    widths = compute_widths(issues, project_context=False)
    lines = [
        format_issue_line(
            issue,
            porcelain=False,
            widths=widths,
            project_context=False,
            configuration=configuration,
            use_color=True,
        )
        for issue in issues
    ]
    context.formatted_output = "\n".join(lines)


@when('I format the list line for issue "{identifier}"')
def when_format_list_line_for_issue(context: object, identifier: str) -> None:
    if not hasattr(context, "original_no_color"):
        context.original_no_color = os.environ.get("NO_COLOR")
    os.environ.pop("NO_COLOR", None)
    project_dir = load_project_directory(context)
    configuration = load_project_configuration(get_configuration_path(project_dir))
    issue = read_issue_file(project_dir, identifier)
    widths = compute_widths([issue], project_context=False)
    context.formatted_output = format_issue_line(
        issue,
        porcelain=False,
        widths=widths,
        project_context=False,
        configuration=configuration,
        use_color=True,
    )


@when('I format the list line for issue "{identifier}" with NO_COLOR set')
def when_format_list_line_for_issue_no_color(context: object, identifier: str) -> None:
    if not hasattr(context, "original_no_color"):
        context.original_no_color = os.environ.get("NO_COLOR")
    os.environ["NO_COLOR"] = "1"
    project_dir = load_project_directory(context)
    configuration = load_project_configuration(get_configuration_path(project_dir))
    issue = read_issue_file(project_dir, identifier)
    widths = compute_widths([issue], project_context=False)
    context.formatted_output = format_issue_line(
        issue,
        porcelain=False,
        widths=widths,
        project_context=False,
        configuration=configuration,
    )


@then("each formatted line should contain ANSI color codes")
def then_each_formatted_line_contains_ansi(context: object) -> None:
    output = getattr(context, "formatted_output", "")
    lines = [line for line in output.splitlines() if line.strip()]
    assert lines, "no formatted lines"
    assert all("\x1b[" in line for line in lines)


@then("the formatted output should contain no ANSI color codes")
def then_formatted_output_has_no_ansi(context: object) -> None:
    output = getattr(context, "formatted_output", "")
    lines = [line for line in output.splitlines() if line.strip()]
    assert lines, "no formatted lines"
    assert all("\x1b[" not in line for line in lines)

File: rust/features/steps/list_format_steps.rs
use std::fs;
use std::path::PathBuf;

use chrono::{TimeZone, Utc};
use cucumber::{given, then, when};

use taskulus::config_loader::load_project_configuration;
use taskulus::file_io::load_project_directory;
use taskulus::issue_line::{compute_widths, format_issue_line};
use taskulus::models::IssueData;

use crate::step_definitions::initialization_steps::TaskulusWorld;

fn load_project_dir(world: &TaskulusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn build_issue(
    identifier: &str,
    title: &str,
    issue_type: &str,
    status: &str,
    parent: Option<&str>,
    priority: i32,
) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: title.to_string(),
        description: "".to_string(),
        issue_type: issue_type.to_string(),
        status: status.to_string(),
        priority,
        assignee: None,
        creator: None,
        parent: parent.map(|value| value.to_string()),
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

fn write_issue_file(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("issues for list color coverage exist")]
fn given_issues_for_list_color_coverage(world: &mut TaskulusWorld) {
    let project_dir = load_project_dir(world);
    let issues = vec![
        build_issue("tsk-line-epic", "Epic", "epic", "open", None, 0),
        build_issue(
            "tsk-line-task",
            "Task",
            "task",
            "in_progress",
            Some("tsk-line-epic"),
            1,
        ),
        build_issue("tsk-line-bug", "Bug", "bug", "blocked", None, 2),
        build_issue("tsk-line-story", "Story", "story", "closed", None, 3),
        build_issue("tsk-line-chore", "Chore", "chore", "deferred", None, 4),
        build_issue(
            "tsk-line-initiative",
            "Initiative",
            "initiative",
            "unknown",
            None,
            9,
        ),
        build_issue(
            "tsk-line-sub",
            "Sub",
            "sub-task",
            "open",
            Some("tsk-line-epic"),
            2,
        ),
        build_issue("tsk-line-event", "Event", "event", "open", None, 2),
        build_issue("tsk-line-unknown", "Unknown", "mystery", "open", None, 2),
    ];
    for issue in issues {
        write_issue_file(&project_dir, &issue);
    }
}

#[when("I format list lines for color coverage")]
fn when_format_list_lines_for_color_coverage(world: &mut TaskulusWorld) {
    let original_no_color = std::env::var("NO_COLOR").ok();
    std::env::remove_var("NO_COLOR");

    let project_dir = load_project_dir(world);
    let config_path = project_dir
        .parent()
        .unwrap_or(&project_dir)
        .join(".taskulus.yml");
    let configuration = if config_path.exists() {
        Some(load_project_configuration(&config_path).expect("load configuration"))
    } else {
        None
    };
    let mut issues = Vec::new();
    for entry in fs::read_dir(project_dir.join("issues")).expect("read issues dir") {
        let entry = entry.expect("issue entry");
        let contents = fs::read_to_string(entry.path()).expect("read issue");
        let issue: IssueData = serde_json::from_str(&contents).expect("parse issue");
        issues.push(issue);
    }
    let widths = compute_widths(&issues, false);
    let mut lines = Vec::new();
    for issue in &issues {
        lines.push(format_issue_line(
            issue,
            Some(&widths),
            false,
            false,
            configuration.as_ref(),
        ));
        lines.push(format_issue_line(issue, Some(&widths), false, false, None));
    }
    world.formatted_output = Some(lines.join("\n"));

    match original_no_color {
        Some(value) => std::env::set_var("NO_COLOR", value),
        None => std::env::remove_var("NO_COLOR"),
    }
}

#[when(expr = "I format the list line for issue {string}")]
fn when_format_list_line_for_issue(world: &mut TaskulusWorld, identifier: String) {
    let original_no_color = std::env::var("NO_COLOR").ok();
    std::env::remove_var("NO_COLOR");

    let project_dir = load_project_dir(world);
    let config_path = project_dir
        .parent()
        .unwrap_or(&project_dir)
        .join(".taskulus.yml");
    let configuration = if config_path.exists() {
        Some(load_project_configuration(&config_path).expect("load configuration"))
    } else {
        None
    };
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    let issue: IssueData = serde_json::from_str(&contents).expect("parse issue");
    let widths = compute_widths(std::slice::from_ref(&issue), false);
    let line = format_issue_line(&issue, Some(&widths), false, false, configuration.as_ref());
    world.formatted_output = Some(line);

    match original_no_color {
        Some(value) => std::env::set_var("NO_COLOR", value),
        None => std::env::remove_var("NO_COLOR"),
    }
}

#[then("each formatted line should contain ANSI color codes")]
fn then_each_formatted_line_contains_ansi(world: &mut TaskulusWorld) {
    let output = world.formatted_output.as_deref().unwrap_or("");
    let lines: Vec<&str> = output
        .lines()
        .filter(|line| !line.trim().is_empty())
        .collect();
    assert!(!lines.is_empty(), "no formatted lines");
    assert!(lines.iter().all(|line| line.contains("\u{1b}[")));
}
